packages/common/src/hooks/useHarmonyField.ts

```diff
@@ -0,0 +1,262 @@
+
+/** Extracts the event target's value from the event for both React and ReactNative. */
+const getEventTargetValue = (
+  e: React.ChangeEvent<any> | React.FocusEvent<any>
+) => e.target?.value ?? e.currentTarget?.value ?? (e as any).nativeEvent?.text
+
+export type UseHarmonyFieldProps<Value> = FieldHookConfig<Value> & {
+  /** Function to transform the input value on change, eg. a function to trim whitespace */
+  transformValueOnChange?: (value: string) => string
+  /** Function to transform the input value on blur, eg. a function to trim whitespace */
+  transformValueOnBlur?: (value: string) => string
+  /**
+   * Delays onChange validation errors from showing. Useful for improving the
+   * UX by not showing users the errors as they're typing into the form input.
+   *
+   * Note: This won't globally debounce validation for the form, just for this
+   * field. Do not use this for eg. debouncing validation API requests unless y
+   * you use it for _every_ field of the form (or other fields don't validate
+   * on change). Otherwise changes to other fields will trigger non-debounced
+   * validation requests anyway.
+   *
+   * Requires validateOnChange to be true (either for the form or this field).
+   *
+   * The default of 0 means errors will be hidden until the user unfocuses.
+   *
+   * @default 0
+   */
+  debouncedIdleMs?: number
+  /**
+   * Whether to show errors prior to the first submission attempt.
+   * @default false
+   */
+  showErrorBeforeSubmit?: boolean
+  /**
+   * Whether to validate this particular field on change or not.
+   * If present, overrides the form's validateFieldOnChange.
+   * @default undefined
+   */
+  validateFieldOnChange?: boolean
+}
+
+/**
+ * Wrapper around Formik's {@link useFormikField useField} that modifies the
+ * returned results for Harmony compatible behavior, including the following
+ * configurable behaviors:
+ * - Only show errors after a submission has been attempted
+ * - Apply transformations to the value on blur or change
+ * - Debounce validations and error displays to lighten the validation
+ *   workload or hide the errors while the user is still typing.
+ * - Allow overriding the Formik context's "validateOnChange"
+ *
+ * Most of the extra functionality is reserved for fields that users type into,
+ * eg. text, email, password inputs.
+ * Unlike the useField hook from Formik, this implementation relies much less
+ * on the event target in onChange and onBlur, instead using the passed in
+ * field name path. This makes usage with ReactNative more straightforward.
+ *
+ * @param propsOrFieldName the field's name, or the full props for the hook.
+ * @example
+ * const MyComponent = () => {
+ *    const [emailField, { error }] = useHarmonyField({ name: 'email', label: 'Email' })
+ *    return <TextInput {...field} error={!!error} helperText={error} />
+ * }
+ */
+export const useHarmonyField = <Value = any>(
+  propsOrFieldName: string | UseHarmonyFieldProps<Value>
```

Comment:
nit: seems a bit convoluted to attempt to squeeze in `string` as an option here. I get why it would be nice to just do `useHarmonyField("email")` but I think it's probably just as easy to do `useHarmonyField({name: "email"})` and that way its a bit more standardized

Reply:
yeah, given we're likely doing this in field components anyway it's somewhat moot. I did want to keep parity with the formik hook though, not sure if that's worth it